Number_types/include/CGAL/Quotient.h

```diff
@@ -685,12 +783,239 @@ template < class NT > class Real_embeddable_traits_quotient_base< Quotient<NT> >
+          const bool is_den_pos = CGAL::is_positive(x.den);
+          if (!is_num_pos && !is_den_pos) {
+            x.num = -x.num;
+            x.den = -x.den;
+          } else if (!is_num_pos && is_den_pos) {
+            change_sign = true;
+            x.num = -x.num;
+          } else if (is_num_pos && !is_den_pos) {
+            change_sign = true;
+            x.den = -x.den;
+          }
+          CGAL_assertion(x.num > 0 && x.den > 0);
+
+          const auto num = get_interval_exp(x.num);
+          const auto den = get_interval_exp(x.den);
+
+          const Interval_nt<> div = num.first / den.first;
+          const int64_t e = num.second - den.second;
+          std::tie(l, u) = ldexp(div, e).pair();
+
+          if (change_sign) {
+            const double t = l;
+            l = -u;
+            u = -t;
+          }
+
+          CGAL_assertion(are_correct_bounds(l, u, input));
+          return std::make_pair(l, u);
+        }
+
+        std::pair<double, double> get_interval_as_boost( Type x ) const {
+
+          CGAL_assertion_code(const Type input = x);
+          double l = 0.0, u = 0.0;
+          if (CGAL::is_zero(x.num)) { // return [0.0, 0.0]
+            CGAL_assertion(are_correct_bounds(l, u, input));
+            return std::make_pair(l, u);
+          }
+          CGAL_assertion(x.num != 0);
+          CGAL_assertion(x.den != 0);
+
+          // Handle signs.
+          bool change_sign = false;
+          const bool is_num_pos = CGAL::is_positive(x.num);
+          const bool is_den_pos = CGAL::is_positive(x.den);
+          if (!is_num_pos && !is_den_pos) {
+            x.num = -x.num;
+            x.den = -x.den;
+          } else if (!is_num_pos && is_den_pos) {
+            change_sign = true;
+            x.num = -x.num;
+          } else if (is_num_pos && !is_den_pos) {
+            change_sign = true;
+            x.den = -x.den;
+          }
+          CGAL_assertion(x.num > 0 && x.den > 0);
+
+          const int64_t num_dbl_digits = std::numeric_limits<double>::digits - 1;
+          const int64_t msb_num = static_cast<int64_t>(msb(x.num));
+          const int64_t msb_den = static_cast<int64_t>(msb(x.den));
+          const int64_t msb_diff = msb_num - msb_den;
+          const int64_t shift = num_dbl_digits - msb_diff;
+
+          if (shift > 0) {
+            CGAL_assertion(msb_diff < num_dbl_digits);
+            x.num <<= shift;
+          } else if (shift < 0) {
+            CGAL_assertion(msb_diff > num_dbl_digits);
+            x.den <<= boost::multiprecision::detail::unsigned_abs(shift);
+          }
+          CGAL_assertion(num_dbl_digits ==
+            static_cast<int64_t>(msb(x.num)) - static_cast<int64_t>(msb(x.den)));
+
+          NT q, r;
+          boost::multiprecision::divide_qr(x.num, x.den, q, r);
+          const int64_t q_bits = static_cast<int64_t>(msb(q));
+          CGAL_assertion(q_bits == num_dbl_digits || r != 0 /* when q_bit = num_dbl_digits - 1 */ );
```

Comment:
In that code, `num_dbl_digits` is always compare to a difference of msb, and I think the value could be `53`, with a small change: add `+ 1` to `q_bits`:
```c++
const int64_t q_bits = static_cast<int64_t>(msb(q)) + 1;
```

Reply:
In fact, `q_bits` is not even used. I put it inside `CGAL_assertion_code` because it is used only inside assertions. So changing it does not change anything. However, if I change 52 to 53, we end up with the zero remainder and then everything goes wrong. This will also require changing the conditions for setting `p` and `q` that is what we tried to do previously and it never worked.

Reply:
This part of the code is changed quite a lot now and it seems to guarantee the tight interval so I think it is fixed.